dmaws/rds.py

```diff
@@ -0,0 +1,206 @@
+                self.log("Waiting while deleting {}".format(name))
+                time.sleep(20)
+                target.update()
+        except boto.exception.BotoServerError as e:
+            if e.status != 404:
+                raise
+
+
+class RDSPostgresClient(object):
+    @staticmethod
+    def from_boto(instance, database, user, password, logger=None):
+        return RDSPostgresClient(
+            instance.endpoint[0], instance.endpoint[1],
+            database, user, password, logger)
+
+    @staticmethod
+    def from_url(url, user, password, logger=None):
+        host, database = url.split('/')
+        host, port = host.split(':')
+        return RDSPostgresClient(host, port, database, user, password, logger)
+
+    def __init__(self, host, port, database, user, password, logger=None):
+        self.db_params = dict(
+            host=host,
+            port=port,
+            database=database,
+            user=user,
+            password=password,
+        )
+        self._cursor = None
+
+        self.log = logger or (lambda *args, **kwargs: None)
+
+    @property
+    def cursor(self):
+        if not self._cursor:
+            self._connection = psycopg2.connect(**self.db_params)
+            self._cursor = self._connection.cursor()
+
+        return self._cursor
+
+    @property
+    def db_path(self):
+        return "postgres://{user}:{password}@{host}:{port}/{database}".format(**self.db_params)
+
+    def execute(self, sql):
+        return self.cursor.execute(sql)
+
+    def commit(self):
+        self._connection.commit()
+
+    def close(self):
+        if self._cursor:
+            self._cursor.close()
+            self._connection.close()
+
+    def dump(self, output_path):
+        utils.run_cmd([
+            'pg_dump', '-cb', '-d', self.db_path, '-f', output_path
+        ], logger=self.log, ignore_errors=True)
+
+    def load(self, export_path):
+        utils.run_cmd([
+            'psql', '-d', self.db_path, 'f', export_path
+        ], logger=self.log, ignore_errors=True)
+
+    def get_open_framework_ids(self):
+        self.cursor.execute("SELECT id FROM frameworks WHERE status = 'open'")
+        return tuple(framework[0] for framework in self.cursor.fetchall())
+
+    def clean_database_for_staging(self):
+        self.log("Update users")
+        self.cursor.execute(
+            "UPDATE users SET name = %s, email_address = id || '@example.com'",
+            ("Example User",))
+
+        open_framework_ids = self.get_open_framework_ids()
+        self.log("Delete draft services")
+        self.cursor.execute(
+            "DELETE FROM draft_services WHERE framework_id IN %s",
+            (open_framework_ids,))
+        self.log("Delete supplier frameworks")
+        self.cursor.execute(
+            "DELETE FROM supplier_frameworks WHERE framework_id IN %s",
+            (open_framework_ids,))
+
+        # Fix suppliers with services but no supplier_framework
```

Comment:
The data on preview will get replaced though, what matters is that `supplier_frameworks` records for all live frameworks are not affected by the deletion of open frameworks. If that's the case then whatever live supplier frameworks exist (or not) in production should be fine for staging and preview (assuming production data is correct, otherwise we need to fix that first).